File: identity_iota_core/src/rebased/iota/move_calls/asset/create.rs
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use iota_interaction::types::base_types::ObjectID;
use iota_interaction::types::programmable_transaction_builder::ProgrammableTransactionBuilder;
use iota_interaction::types::transaction::Command;
use iota_interaction::types::transaction::ProgrammableMoveCall;
use iota_interaction::types::transaction::ProgrammableTransaction;
use iota_interaction::ident_str;
use serde::Serialize;

use iota_interaction::MoveType;
use crate::rebased::Error;
use super::try_to_argument;

pub(crate) fn new<T: Serialize + MoveType>(
  inner: T,
  mutable: bool,
  transferable: bool,
  deletable: bool,
  package: ObjectID,
) -> Result<ProgrammableTransaction, Error> {
  let mut ptb = ProgrammableTransactionBuilder::new();
  let inner = try_to_argument(&inner, &mut ptb, package)?;
  let mutable = ptb.pure(mutable).map_err(|e| Error::InvalidArgument(e.to_string()))?;
  let transferable = ptb
    .pure(transferable)
    .map_err(|e| Error::InvalidArgument(e.to_string()))?;
  let deletable = ptb.pure(deletable).map_err(|e| Error::InvalidArgument(e.to_string()))?;

  ptb.command(Command::MoveCall(Box::new(ProgrammableMoveCall {
    package,
    module: ident_str!("asset").into(),
    function: ident_str!("new_with_config").into(),
    type_arguments: vec![T::move_type(package)],
    arguments: vec![inner, mutable, transferable, deletable],
  })));

  Ok(ptb.finish())
}

File: bindings/wasm/iota_interaction_ts/lib/move_calls/asset/create.ts
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { Transaction } from "@iota/iota.js/transactions";

export function create(
    inner_bytes: Uint8Array,
    inner_type: string,
    mutable: boolean,
    transferable: boolean,
    deletable: boolean,
    packageId: string,
): Promise<Uint8Array> {
    const tx = new Transaction();
    const inner_arg = tx.pure(inner_bytes);
    const mutableArg = tx.pure.bool(mutable);
    const transferableArg = tx.pure.bool(transferable);
    const deletableArg = tx.pure.bool(deletable);

    tx.moveCall({
        target: `${packageId}::asset::new_with_config`,
        typeArguments: [inner_type],
        arguments: [inner_arg, mutableArg, transferableArg, deletableArg],
    });

    return tx.build();
}
